File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        // Base case: if the root is null or it’s a leaf node, simply return the root
        if (root == null || (root.left == null && root.right == null)) {
            return root;
        }

        // Initialize a stack to store nodes at the current level, starting with the root
        Stack<Node> stack = new Stack<>();
        stack.add(root);
        
        // `current` is used to link nodes horizontally at each level
        Node current = root;
        
        // `nodeleft` tracks the leftmost node at each level, used for moving down to the next level
        Node nodeleft = current.left;

        // Outer loop to process each level until all nodes are connected
        while (!stack.isEmpty()) {
            // Stack to hold nodes for the next level
            Stack<Node> nextlevel = new Stack<>();
            
            // Loop through each node in the current level
            for (Node node : stack) {
                // If the node has children, add them to the `nextlevel` stack
                if (node.left != null) {
                    nextlevel.add(node.left);
                    nextlevel.add(node.right);
                }
            }

            // Connect each node in the current level stack to its next node
            for (int i = 1; i < stack.size(); i++) {
                current.next = stack.get(i);  // Link current node to the next node in the level
                current = current.next;       // Move `current` to the next node
            }

            // Reset `current` to the leftmost node of the next level
            current = nodeleft;

            // Move `nodeleft` to the leftmost child of the next level, if it exists
            if (nodeleft.left != null) {
                nodeleft = nodeleft.left;
            }

            // Set `stack` to `nextlevel` to process the next level in the next iteration
            stack = nextlevel;
        }

        // Return the modified tree with `next` pointers connected
        return root;
    }
}

File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        # Base case: if root is None or it's a leaf node (no children), return the root
        if not root or (not root.left and not root.right):
            return root

        # Initialize stack to keep track of nodes at the current level, starting with the root node
        stack = [root]
        
        # `current` keeps track of the node to set `next` pointers for, starting with the root
        current = root
        
        # `node` is used to track the leftmost node of the next level
        node = current.left
         
        # Process each level until there are no more levels to traverse
        while stack:
            # `nextlevel` list to hold nodes at the next level
            nextlevel = []
            
            # Traverse each node in the current level stack
            for tree in stack:
                # If the current node has children, add them to `nextlevel`
                if tree.left:
                    nextlevel.append(tree.left)
                    nextlevel.append(tree.right)
            
            # Connect nodes horizontally in the current level
            for i in range(1, len(stack)):
                # Set `next` of the `current` node to the next node in the list
                current.next = stack[i]
                
                # Move `current` to the next node in the same level
                current = current.next
            
            # Reset `current` to `node`, which is the leftmost node of the next level
            current = node
            
            # Update `node` to the left child of the current `node` for the next level
            if node.left:
                node = node.left
            
            # Set `stack` to `nextlevel` to process the next level in the next iteration
            stack = nextlevel

        # Return the modified tree with connected `next` pointers
        return root
            


        